src/circuit/gadget/sinsemilla/chip/hash_to_point.rs

```diff
@@ -0,0 +1,369 @@
+use super::super::SinsemillaInstructions;
+use super::{get_s_by_idx, CellValue, EccPoint, SinsemillaChip, Var};
+use crate::circuit::gadget::utilities::copy;
+use crate::primitives::sinsemilla::lebs2ip_k;
+use halo2::{
+    arithmetic::{CurveAffine, FieldExt},
+    circuit::{Chip, Region},
+    plonk::Error,
+};
+
+use ff::{Field, PrimeFieldBits};
+use group::Curve;
+use pasta_curves::pallas;
+
+use std::ops::Deref;
+
+#[cfg(test)]
+use group::prime::PrimeCurveAffine;
+
+impl<const K: usize, const MAX_WORDS: usize> SinsemillaChip<K, MAX_WORDS> {
+    #[allow(non_snake_case)]
+    #[allow(clippy::type_complexity)]
+    pub(super) fn hash_message(
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        Q: pallas::Affine,
+        message: &<Self as SinsemillaInstructions<pallas::Affine, K, MAX_WORDS>>::Message,
+    ) -> Result<(EccPoint, Vec<CellValue<pallas::Base>>), Error> {
+        let config = self.config().clone();
+
+        // Get the `x`- and `y`-coordinates of the starting `Q` base.
+        let x_q = *Q.coordinates().unwrap().x();
+        let y_q = *Q.coordinates().unwrap().y();
+
+        // Assign the `x`-coordinate of the starting `Q` base.
+        let x_q_cell = region.assign_advice(|| "x_q", config.x_a, 0, || Ok(x_q))?;
+
+        // Initialize the accumulator to `Q`.
+        let mut x_a: X<pallas::Base> = CellValue::new(x_q_cell, Some(x_q)).into();
+        let mut y_a: Y<pallas::Base> = Some(y_q).into();
+        let mut zs_sum: Vec<CellValue<pallas::Base>> = Vec::new();
+
+        // Hash each piece in the message.
+        let mut offset = 0;
+        for piece in message.0.iter() {
+            // The value of the hash after this piece is processed.
+            let (x, y, zs) = self.hash_piece(region, offset, piece, x_a, y_a)?;
+
+            // Since each message word takes one row to process, we increase
+            // the offset by `piece.num_words` on each iteration.
+            offset += piece.num_words() + 1;
+
+            // Update the accumulator to the latest hash value.
+            x_a = x;
+            y_a = y;
+            zs_sum.extend_from_slice(&zs);
```

Comment:
Now that I understand the `zs` returned from `hash_piece` is intended to be disjoint for each piece, this should not be extending `zs_sum` (because that leaves us with a single `Vec` containing all of the pieces' running sums, which then need to be correctly split apart again). Instead, make `zs_sum` a `Vec<Vec<CellValue<_>>>` so the caller gets each running sum independently.